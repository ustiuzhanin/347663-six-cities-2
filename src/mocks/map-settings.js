export const mapSettings = {
  icon: {
    url: `/img/pin.svg`,
    size: [30, 30]
  },
  activeIcon: {
    url: `/img/pin-active.svg`,
    size: [30, 30]
  },
  defaultCity: [52.37454, 4.897976],
  defaultZoom: 13
};
