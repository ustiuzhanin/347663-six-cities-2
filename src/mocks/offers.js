export const offers = [
  {
    id: 1,
    src: `img/room.jpg`,
    price: `80`,
    rating: `70`,
    name: `Beautiful & luxurious apartment at great location`,
    type: `Private room`
  },
  {
    id: 2,
    src: `img/room.jpg`,
    price: `120`,
    rating: `40`,
    name: `Wood and stone place`,
    type: `House`
  },
  {
    id: 3,
    src: `img/room.jpg`,
    price: `20`,
    rating: `100`,
    name: `Canal View Prinsengracht`,
    type: `Private room`
  },
  {
    id: 4,
    src: `img/room.jpg`,
    price: `60`,
    rating: `50`,
    name: `Nice, cozy, warm big bed apartment`,
    type: `Apartment`
  }
];
