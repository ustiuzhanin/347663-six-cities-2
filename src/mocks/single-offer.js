export const offer = {
  city: {
    _id: "6005f36637751d35c8503011",
    name: `Dusseldorf`,
    location: { latitude: 51.225402, longitude: 6.776314, zoom: 13 },
  },
  preview_image: `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg`,
  images: [
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/19.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/8.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg`,
  ],
  title: `Canal View Prinsengracht`,
  is_premium: false,
  rating: 3.1,
  type: `room`,
  bedrooms: 1,
  max_adults: 2,
  price: "101",
  goods: [`Breakfast`, `Laptop friendly workspace`],
  host: "6005f36637751d35c8503011",
  description: `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  location: {
    latitude: 51.248402000000006,
    longitude: 6.763314,
    zoom: 16,
  },
  _id: "6005f36637751d35c8503011",
};
export const offerWithHost = {
  city: {
    _id: "6005f36637751d35c8503011",
    name: `Dusseldorf`,
    location: { latitude: 51.225402, longitude: 6.776314, zoom: 13 },
  },
  preview_image: `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg`,
  images: [
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/19.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/8.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg`,
    `https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg`,
  ],
  title: `Canal View Prinsengracht`,
  is_premium: false,
  rating: 3.1,
  type: `room`,
  bedrooms: 1,
  max_adults: 2,
  price: "101",
  goods: [`Breakfast`, `Laptop friendly workspace`],
  host: {
    _id: "6005f36637751d35c8503011",
    avatar_url: "/img/avatar.svg",
    bookmarks: ["5fee3aaa54b627023d7f94b7"],
    email: "test1@test.com",
    is_pro: false,
    name: "test",
    offers: [],
  },
  description: `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  location: {
    latitude: 51.248402000000006,
    longitude: 6.763314,
    zoom: 16,
  },
  _id: "6005f36637751d35c8503011",
};
