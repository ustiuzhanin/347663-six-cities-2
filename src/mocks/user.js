export const user = {
  _id: "6005f36637751d35c8503011",
  avatar_url: "/img/avatar.svg",
  bookmarks: ["5fee3aaa54b627023d7f94b7"],
  email: "test1@test.com",
  is_pro: false,
  name: "test",
  offers: [],
};
