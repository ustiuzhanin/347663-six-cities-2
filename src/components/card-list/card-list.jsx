import React, {useEffect, useState} from 'react';
import {connect} from 'react-redux';
import {ActionCreator} from '../../reducer/offers/offers';
import PropTypes from 'prop-types';

import Card from '../card/card.jsx';

/* eslint-disable camelcase*/

const SortingType = {
  POPULAR: 'popular',
  RATING: 'price-raitig',
  PRICE_LOW_TO_HIGH: 'price-low',
  PRICE_HIGH_TO_LOW: 'price-high',
};

const CardList = (props) => {
  const {
    activeCity,
    offers,
    addActiveCityOffers,
    listOfOffers,
    onCardHeaderClick,
    resetOffers,
    sorting,
    changeSortingType,
  } = props;

  const [isSortPopupOpen, setIsSortPopupOpen] = useState(false);

  useEffect(() => {
    const closeSelect = () => {
      setIsSortPopupOpen(false);

      document.removeEventListener('click', closeSelect);
    };

    if (isSortPopupOpen) {
      document.addEventListener('click', closeSelect);
    }
  });

  useEffect(() => {
    const activeCityOffers = [];
    resetOffers();

    offers.filter((offer) => {
      const {city} = offer;

      if (city.name === activeCity) {
        activeCityOffers.push(offer);
      }
    });

    addActiveCityOffers(activeCityOffers);
  }, [activeCity]);

  const sortingClickHandler = () => {
    setIsSortPopupOpen(!isSortPopupOpen);
  };

  const optionClickHandler = (e) => {
    if (e.target.id !== sorting.type) {
      setIsSortPopupOpen(false);

      changeSortingType({type: e.target.id, text: e.target.innerText});
    }
  };

  const sortOffers = (sort, arr) => {
    switch (sort.type) {
      case SortingType.POPULAR:
        arr.sort((a, b) => {
          return a.id - b.id;
        });
        break;
      case SortingType.RATING:
        arr.sort((a, b) => {
          return b.rating - a.rating;
        });
        break;
      case SortingType.PRICE_LOW_TO_HIGH:
        arr.sort((a, b) => {
          return a.price - b.price;
        });
        break;
      case SortingType.PRICE_HIGH_TO_LOW:
        arr.sort((a, b) => {
          return b.price - a.price;
        });
        break;
    }
    return arr;
  };

  return (
    <section className='cities__places places'>
      <h2 className='visually-hidden'>Places</h2>
      <b className='places__found'>
        {listOfOffers.length} places to stay in {activeCity}
      </b>
      <form className='places__sorting' action='#' method='get'>
        <span className='places__sorting-caption'>Sort by</span>
        <span
          className='places__sorting-type'
          tabIndex='0'
          onClick={sortingClickHandler}
        >
          {sorting.text}
          <svg className='places__sorting-arrow' width='7' height='4'>
            <use xlinkHref='#icon-arrow-select'></use>
          </svg>
        </span>
        <ul
          className={`places__options places__options--custom ${
            isSortPopupOpen ? 'places__options--opened' : null
          }`}
        >
          <li
            className='places__option'
            tabIndex='0'
            onClick={(evt) => optionClickHandler(evt)}
            id='popular'
          >
            Popular
          </li>
          <li
            className='places__option'
            tabIndex='0'
            onClick={(evt) => optionClickHandler(evt)}
            id='price-low'
          >
            Price: low to high
          </li>
          <li
            className='places__option'
            tabIndex='0'
            onClick={(evt) => optionClickHandler(evt)}
            id='price-high'
          >
            Price: high to low
          </li>
          <li
            className='places__option'
            tabIndex='0'
            onClick={(evt) => optionClickHandler(evt)}
            id='price-raitig'
          >
            Top rated first
          </li>
        </ul>

        {/* <select className='places__sorting-type' id='places-sorting'>
          <option className='places__option' value='popular' selected=''>
            Popular
          </option>
          <option className='places__option' value='to-high'>
            Price: low to high
          </option>
          <option className='places__option' value='to-low'>
            Price: high to low
          </option>
          <option className='places__option' value='top-rated'>
            Top rated first
          </option>
        </select> */}
      </form>
      <div className='cities__places-list places__list tabs__content'>
        {sortOffers(sorting, listOfOffers).map((card) => (
          <Card
            key={card.id}
            card={card}
            onCardHeaderClick={onCardHeaderClick}
          />
        ))}
      </div>
    </section>
  );
};

CardList.propTypes = {
  offers: PropTypes.arrayOf(
    PropTypes.shape({
      price: PropTypes.number.isRequired,
      rating: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      preview_image: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
  listOfOffers: PropTypes.arrayOf(
    PropTypes.shape({
      price: PropTypes.number.isRequired,
      rating: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      preview_image: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
  activeCity: PropTypes.string.isRequired,
  addActiveCityOffers: PropTypes.func.isRequired,
  onCardHeaderClick: PropTypes.func.isRequired,
  resetOffers: PropTypes.func.isRequired,
  sorting: PropTypes.shape({
    type: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  changeSortingType: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    activeCity: state.cityList.activeCity,
    listOfOffers: state.offers.listOfOffers,
    sorting: state.offers.sorting,
  });

const mapDispatchToProps = (dispatch) => ({
  addActiveCityOffers: (activeOffers) => {
    dispatch(ActionCreator.addActiveCityOffers(activeOffers));
  },
  changeSortingType: (type) => {
    dispatch(ActionCreator.changeSorting(type));
  },
  resetOffers: () => {
    dispatch(ActionCreator.resetOffersList());
  },
});

export {CardList};

export default connect(mapStateToProps, mapDispatchToProps)(CardList);

/* eslint-enable camelcase*/
