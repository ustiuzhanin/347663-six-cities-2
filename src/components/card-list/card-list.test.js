import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import { CardList } from "./card-list.jsx";

import { offer as MockOffer } from "../../mocks/single-offer";

test(`CardList's snapshot`, () => {
  const renderer = new ShallowRenderer();

  renderer.render(
    <CardList
      activeCity={`Amsterdam`}
      changeSortingType={jest.fn()}
      loadCityOffers={jest.fn()}
      sorting={{ type: "popular", text: "Popular" }}
      cityOffers={[MockOffer]}
    />
  );

  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});
