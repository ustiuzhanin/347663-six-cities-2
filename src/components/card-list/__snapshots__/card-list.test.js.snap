// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList's snapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <CardList
    activeCity="Amsterdam"
    addActiveCityOffers={[MockFunction]}
    changeSortingType={[MockFunction]}
    listOfOffers={
      Array [
        Object {
          "bedrooms": 1,
          "city": Object {
            "location": Object {
              "latitude": 51.225402,
              "longitude": 6.776314,
              "zoom": 13,
            },
            "name": "Dusseldorf",
          },
          "description": "This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.",
          "goods": Array [
            "Breakfast",
            "Laptop friendly workspace",
          ],
          "host": Object {
            "avatar_url": "img/avatar-angelina.jpg",
            "id": 25,
            "is_pro": true,
            "name": "Angelina",
          },
          "id": 1,
          "images": Array [
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/19.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/20.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/8.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg",
          ],
          "is_favorite": false,
          "is_premium": false,
          "location": Object {
            "latitude": 51.248402000000006,
            "longitude": 6.763314,
            "zoom": 16,
          },
          "max_adults": 2,
          "preview_image": "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg",
          "price": 101,
          "rating": 3.1,
          "title": "Canal View Prinsengracht",
          "type": "room",
        },
      ]
    }
    offers={
      Array [
        Object {
          "bedrooms": 1,
          "city": Object {
            "location": Object {
              "latitude": 51.225402,
              "longitude": 6.776314,
              "zoom": 13,
            },
            "name": "Dusseldorf",
          },
          "description": "This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.",
          "goods": Array [
            "Breakfast",
            "Laptop friendly workspace",
          ],
          "host": Object {
            "avatar_url": "img/avatar-angelina.jpg",
            "id": 25,
            "is_pro": true,
            "name": "Angelina",
          },
          "id": 1,
          "images": Array [
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/19.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/20.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/8.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg",
            "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg",
          ],
          "is_favorite": false,
          "is_premium": false,
          "location": Object {
            "latitude": 51.248402000000006,
            "longitude": 6.763314,
            "zoom": 16,
          },
          "max_adults": 2,
          "preview_image": "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/9.jpg",
          "price": 101,
          "rating": 3.1,
          "title": "Canal View Prinsengracht",
          "type": "room",
        },
      ]
    }
    onCardHeaderClick={[MockFunction]}
    resetOffers={[MockFunction]}
    sorting={
      Object {
        "text": "Popular",
        "type": "popular",
      }
    }
  />
</Router>
`;
